### PR TITLE
avm2: Work on AMF3 custom object serialization

### PR DESCRIPTION
**Pull request by Aaron1011**
I still need to write tests for this

### CODE REVIEW

File: core/src/avm2/globals/flash/net.as [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
+    internal var _aliasToClass: Object = {};
+    internal var _classToAlias: Dictionary = new Dictionary();
 
     public native function navigateToURL(request:URLRequest, window:String = null):void;
 
     public function registerClassAlias(name:String, object:Class):void {
-        stub_method("flash.net", "registerClassAlias");
-        this._classLookups[name] = object;
+        if (name == null || object == null) {
+            throw new TypeError("FIXME - what error should I use");
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/net.as:14]:
```
if (name == null) {
    throw new TypeError("Error #2007: Parameter aliasName must be non-null.", 2007);
}
if (object == null) {
    throw new TypeError("Error #2007: Parameter classObject must be non-null.", 2007);
}
```

**Comment by n0samu** [comment]
Testing [Axon](http://i.notdoppler.com/files/axon.swf) and [Catchy Orbit](https://www.newgrounds.com/portal/view/630301/format/flash?emulate=flash), I get a panic when closing the game, and after reopening the game my best score has not been saved:
```
panicked at 'called `Option::unwrap()` on a `None` value', core\src\avm2\amf.rs:84:22
```

File: core/src/avm2/amf.rs [hunk 2]
```diff
@@ -118,31 +121,56 @@ pub fn serialize_value<'gc>(
...
-                            None
-                        },
-                    ))
+                let class = o.instance_of().expect("Missing ClassObject");
+                let qname = QName::new(activation.avm2().flash_net_internal, "_getAliasByClass");
+                let method = activation
+                    .avm2()
+                    .playerglobals_domain
+                    .get_defined_value(activation, qname)
+                    .expect("Failed to lookup flash.net._getAliasByClass");
```
Comment (Lord-McSweeney) [inline core/src/avm2/amf.rs:130]: This feels like a hack- could you instead add `getAliasByClass` into the AVM2 `SystemClasses`?

File: core/src/avm2/globals/flash/utils/byte_array.rs [hunk 3]
```diff
@@ -782,7 +782,6 @@ pub fn read_object<'gc>(
         bytearray.set_position(bytearray.len() - bytes_left);
         return Ok(value);
     }
-
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/utils/byte_array.rs:785]: Unrelated formatting change?

**Comment by MartySVK** [comment]
it looks like [the game](https://github.com/ruffle-rs/ruffle/issues/10888) is no longer soft locked thanks to this PR
Progress is still not saved, but the game has different issues, so maybe that's the problem

**Comment by Dinnerbone** [comment]
Panics for me when I try to play Learn to Fly 3:

```
panicked at core\src\avm2\amf.rs:160:18:
called `Result::unwrap()` on an `Err` value: Error: Requesting a stat from empty stat object

Stack:

339/__wbg_get_imports/imports.wbg.__wbg_new_abda76e883ba8a5f@moz-extension://d3148af3-f0c5-4eb3-a830-e61725407381/dist/339.js:804:21
core::ops::function::FnOnce::call_once{{vtable.shim}}::he99c379da5a31fb9@:wasm-function[9611]:0x9a2be6
std::panicking::rust_panic_with_hook::h3aa054d35a0817d7@:wasm-function[6336]:0x89bf21
core::panicking::panic_fmt::h3e1dd3d08288569e@:wasm-function[9421]:0x99de69
core::result::unwrap_failed::h8b3db0f11171b57b@:wasm-function[7922]:0x918968
ruffle_core::avm2::amf::serialize_value::hb99a4adce0a2bd3b@:wasm-function[817]:0x2bed44
ruffle_core::avm2::amf::recursive_serialize::hbcbbf5470ab600ae@:wasm-function[1194]:0x3f444a
ruffle_core::avm2::amf::serialize_value::hb99a4adce0a2bd3b@:wasm-function[817]:0x2beced
ruffle_core::avm2::amf::recursive_serialize::hbcbbf5470ab600ae@:wasm-function[1194]:0x3f444a
ruffle_core::avm2::amf::serialize_value::hb99a4adce0a2bd3b@:wasm-function[817]:0x2beced
...


```
AVM2 stack:
```
    at flash.utils::ByteArray/flash::utils::ByteArray::writeObject()
    at src.system::SaveLoadModule$/profileToString()
    at src.system::SaveLoadModule$/saveProfileToSharedObject()
    at src::Main$/saveGame()
    at src::Main$/gameToShop()
    at src.gui::Transition/tick()
    at org.osflash.signals::Slot/execute()
    at org.osflash.signals::OnceSignal/dispatch()
    at src::Main/enterFrameTick()
```

**Comment by n0samu** [comment]
Just tested before and after the latest commit, and it fixes the panic Dinnerbone mentioned!

**Comment by n0samu** [comment]
This also fixes #12366 (and the many identical Spil Games issues such as #13703) 🎉

**Review by Dinnerbone — APPROVED** [review]

File: core/src/avm2/amf.rs (hunk 2 above)
Comment (Aaron1011) [reply core/src/avm2/amf.rs:133]: Unfortunately, I can't - this is a package level function, not class-level
